src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -27,6 +27,37 @@ void GcsTaskManager::Stop() {
   }
 }
 
+std::vector<rpc::TaskEvents> GcsTaskManager::GcsTaskManagerStorage::GetTaskEvents(
+    const absl::flat_hash_set<std::string> &task_ids,
+    absl::optional<std::string> job_id) {
+  std::vector<rpc::TaskEvents> results;
+
+  if (!task_ids.empty()) {
+    RAY_CHECK(!job_id.has_value())
+        << "only one of job_id or task_ids should be provided. ";
+    // Get all task events with task ids.
+    for (const auto &task_event : task_events_) {
```

Comment:
This is O(stored tasks), which isn't great for scalability. Do we have an index by task id?